File: 백준/Silver/9012. 괄호/괄호.py
import sys
input=sys.stdin.readline

T = int(input())

for _ in range(T):
    stack = []
    S = input().rstrip()
    
    for i in S:
        if i == '(':
            stack.append(i)
        elif i == ')':
            if stack:
                stack.pop()
            else:
                print('NO')
                break

    # for-else문: for문에서 한번도 break 발생한적 없으면 실행
    else:        
        if not stack:
            print('YES')
        else:
            print('NO')

File: 백준/Silver/9012. 괄호/괄호.cc
#include<bits/stdc++.h>

using namespace std;

int t;
string s;

int main() {
	ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
	
	cin >> t;
	
	while (t--) {
		cin >> s;
		stack<char> st;
		
		for (int i = 0; i < s.size(); i++) {
			char cur = s[i];
			
			if (!st.empty() && st.top() == '(' && cur == ')') {
				st.pop();
			} else {
				st.push(cur);
			}
		}
		
		if (st.empty()) {
			cout << "YES" << "\n";
		} else {
			cout << "NO" << "\n";
		}
	}
		
	return 0;
}